processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -0,0 +1,713 @@
+        .appendByteArray(hasher.hash().asBytes())
+        .build();
+  }
+
+  private static boolean checkSorted(List<?> unsortedValues, ColumnType matchValueType)
+  {
+    final Comparator<Object> comparator = matchValueType.getNullableStrategy();
+    Object prev = null;
+    boolean needsCoerceCheck = true;
+    for (Object o : unsortedValues) {
+      if (needsCoerceCheck && o != null) {
+        Object coerced = coerceValue(o, matchValueType);
+        //noinspection ObjectEquality
+        if (coerced != o) {
+          return false;
+        }
+        needsCoerceCheck = false;
+      }
+      if (prev != null && comparator.compare(prev, o) >= 0) {
+        return false;
+      }
+      prev = o;
+    }
+    return true;
+  }
+
+  @Nullable
+  private static Object coerceValue(@Nullable Object o, ColumnType matchValueType)
+  {
+    if (o == null) {
+      return o;
+    }
+    switch (matchValueType.getType()) {
+      case STRING:
+        return DimensionHandlerUtils.convertObjectToString(o);
+      case LONG:
+        return DimensionHandlerUtils.convertObjectToLong(o);
+      case FLOAT:
+        return DimensionHandlerUtils.convertObjectToFloat(o);
+      case DOUBLE:
+        return DimensionHandlerUtils.convertObjectToDouble(o);
+      default:
+        throw InvalidInput.exception("Unsupported matchValueType[%s]", matchValueType);
+    }
+  }
+
+  private static List<?> sortValues(List<?> unsortedValues, ColumnType matchValueType)
+  {
+    final Object[] array = unsortedValues.toArray(new Object[0]);
+    // check if values need coerced
+    for (int i = 0; i < array.length; i++) {
+      Object coerced = coerceValue(array[i], matchValueType);
+      //noinspection ObjectEquality
+      if (coerced != null && array[i] == coerced) {
+        // assume list is all same type objects...
+        break;
+      }
+      array[i] = coerced;
+    }
+    ObjectArrays.quickSort(array, matchValueType.getNullableStrategy());
+    return Arrays.asList(array);
+  }
+
+  /**
+   * Since jackson might translate longs into ints and such, we use the type comparator to check lists
+   * for {@link #equals(Object)} instead of directly using {@link Objects#equals(Object, Object)}
+   */
+  private static boolean compareValues(List<?> o1, List<?> o2, ColumnType matchValueType)
+  {
+    final NullableTypeStrategy comparator = matchValueType.getNullableStrategy();
+    //noinspection ObjectEquality
+    if (o1 == o2) {
+      return true;
+    }
+    if (o1 == null) {
+      return false;
+    }
+    if (o2 == null) {
+      return false;
+    }
+    final int iter = Math.min(o1.size(), o2.size());
+    for (int i = 0; i < iter; i++) {
+      final int cmp = comparator.compare(o1.get(i), o2.get(i));
+      if (cmp == 0) {
+        continue;
+      }
+      return false;
+    }
+    return o1.size() == o2.size();
+  }
+
+  private static DruidObjectPredicate<String> createStringPredicate(
+      final List sortedValues,
```

Comment:
I usually prefer `List<?>` if it works ok, because it at least keeps the unknownness to the single type param rather than encouraging the dropping of all type params.

Reply:
done